2019-Q4-PyPI/RFP.md

```diff
@@ -0,0 +1,203 @@
+
+#### Summary
+
+* Implementation of [PEP 458 once accepted](https://www.python.org/dev/peps/pep-0458/) to add integration of [The Update Framework](https://theupdateframework.github.io/) to PyPI
+* Development of either a stand alone service or code in the [Warehouse codebase](https://github.com/pypa/warehouse) to create, sign, serve, and handle caching concerns for TUF metadata
+* Development of necessary code in the [Warehouse codebase] to integrate TUF](https://theupdateframework.github.io/) metadata and signing
+
+[Relevant background and context](https://github.com/pypa/warehouse/issues/5247)
+
+#### Specific tasks
+
+* Development of systems for generation and signing of metadata compliant with [TUF](https://theupdateframework.github.io/)
+* Development of systems for serving signed metadata compliant with [TUF](https://theupdateframework.github.io/)
+* Development of systems for appropriately caching and invalidation of caches for metadata service
+* Documentation for Administrators of PyPI for handling key material and interacting with the systems
+* Documentation for PyPI client libraries for sourcing and validating [TUF](https://theupdateframework.github.io/) metadata
+
+**Note**: Implementation of support for TUF signing or verification in the Python packaging toolchain (e.g. setuptools, pip, twine, etc) is not in scope.
+
+### Milestone 2 - Systems for Automated Detection of Malicious Uploads
+
+#### Summary
+
+* Development of infrastructure as either a standalone service or code in the [Warehouse codebase](https://github.com/pypa/warehouse) to automatically screen metadata and uploads to PyPI for malicious content via pluggable checks, reporting results to PyPI
+* Development of necessary code in the [Warehouse codebase](https://github.com/pypa/warehouse) to store results of automated screening for administrator review
+
+[Relevant background and context](https://github.com/pypa/warehouse/issues/5117)
+
+#### Specific tasks
+
+* Develop systems for running automated checks on new Projects, Release, and Release Files.
+* Develop a pluggable system for development and deployment of new checks
+* Implement database models to store check results relating to Projects, Releases, and Release Files from checks
+* Implement Administrator views for reviewing results
+* Documentation process for the development and submission of new checks
+
+**Note**: Surfacing malware detection results to PyPI publishers and users via the Web User Interface or any API is not in scope.
+
+## Estimated Budgets and Caps
+
+**Budgets and caps** are provided to help contractors in preparing their proposals.
+
+Budgets are PSF's estimates, and caps are based on the funding commitment that we have received.
+
+Caps are provided to indicate that we understand that some features may require more funds than our estimated budget. Proposals may go over budget up to the cap for one feature, but fall under budget for another.
```

Comment:
I suggest we open up the possibility that a proposal actually goes way under the cap for 1 task and over the cap for the other task (as long as the proposal still includes a healthy documentation line item, and the total proposal is <= $65,000). Based on my conversations with Justin, I think it's a strong possibility that TUF implementation is actually going to be way faster than we've budgeted for.

Reply:
👍